File: src/org/warp/picalculator/math/rules/ExponentRule3.java
package org.warp.picalculator.math.rules;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.Function;
import org.warp.picalculator.math.functions.Number;
import org.warp.picalculator.math.functions.Power;

/**
 * Exponent rule<br>
 * <b>a^0=1</b>
 * 
 * @author Andrea Cavalli
 *
 */
public class ExponentRule3 {

	public static boolean compare(Function f) {
		final Power fnc = (Power) f;
		if (fnc.getParameter2().equals(new Number(f.getMathContext(), 0))) {
			return true;
		}
		return false;
	}

	public static ObjectArrayList<Function> execute(Function f) throws Error {
		final ObjectArrayList<Function> result = new ObjectArrayList<>();
		result.add(new Number(f.getMathContext(), 1));
		return result;
	}

}

File: src/main/resources/rules/ExponentRule3.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var Number = org.warp.picalculator.math.functions.Number;
var Power = org.warp.picalculator.math.functions.Power;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Exponent rule
 * a^0=1
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "ExponentRule3";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.CALCULATION;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		if (ScriptUtils.instanceOf(f, Power.class)) {
			var fnc = f;
			if (fnc.getParameter2().equals(new Number(f.getMathContext(), 0))) {
				isExecutable = true;
			}
		}
	
		if (isExecutable) {
		var result = new ObjectArrayList();
		result.add(new Number(f.getMathContext(), 1));
		return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));